dev/bots/utils.dart

```diff
@@ -569,9 +570,26 @@ List<T> selectIndexOfTotalSubshard<T>(List<T> tests, {String subshardKey = kSubs
     return <T>[];
   }
 
-  final int testsPerShard = (tests.length / total).ceil();
-  final int start = (index - 1) * testsPerShard;
-  final int end = math.min(index * testsPerShard, tests.length);
+  // While there exists a closed formula figuring out the range of tests this
+  // shard is resposible for, modeling this as a simulation of distributing
+  // items equally into buckets is more intuitive.
+  //
+  // A bucket represents how many tests a shard should be allocated.
+  final List<int> buckets = List<int>.filled(total, 0);
+  // First, allocate an even amount of items to each bucket.
+  for (int i = 0; i < buckets.length; i++) {
+    buckets[i] = (tests.length / total).floor();
+  }
+  // For the N leftover items, put one into each of the first N buckets.
+  final int remainingItems = tests.length % buckets.length;
+  for (int i = 0; i < remainingItems; i++) {
+    buckets[i] += 1;
+  }
+
+  // Lastly, compute the indices of the items in buckets[index].
+  final int numberOfItemsInPreviousBuckets = index == 0 ? 0 : buckets.sublist(0, index - 1).sum;
+  final int start = (numberOfItemsInPreviousBuckets + 1) - 1;
+  final int end = (start + buckets[index - 1]) - 1;
```

Comment:
The formula for directly computing `start` and `end` would look something like this:

```dart
  // The idea here is that we distribute the tests perfectly equally amongst the shards.
  // That is, each shard gets (tests.length / total).floor() tests.
  // However, this leaves N = (tests.length % total) tests left over, so we will
  // put one more test into the first N shards. We call these shards "busy" shards.
  int busyShards = tests.length % total;
  int testsPerBusyShard = (tests.length / total).ceil();
  int testsPerOtherShard = (tests.length / total).floor();
  for (int index = 1; index <= total; index++) {
    int zeroIndexed = index - 1;
    int start = zeroIndexed <= busyShards
        ? zeroIndexed * testsPerBusyShard
        : (busyShards) * testsPerBusyShard +
            (zeroIndexed - busyShards) * testsPerOtherShard;
    int end = start +
        (zeroIndexed <= busyShards ? testsPerBusyShard : testsPerOtherShard) -
        1;

    print('$index: $start, $end');
  }
}
```